# **UNLOCKING REAL ESTATE SUCCESS WITH TIME SERIES MODELLING**

This is a collaborative group project done at the end of Phase 4 of Moringa School's Data Science program. The team members of this group include:

- [Abdideq Adan](https://github.com/AdanAbdideq)
- [Clara Gatambia](https://github.com/claragatambia)
- [Isaack Odera](https://github.com/derak-isaack)
- [Mwiti Mwongo](https://github.com/M13Mwongo)
- [Wilson Mutungu](https://github.com/mutungu)

## 1. BUSINESS UNDERSTANDING

TODO - Answer a sample question like why prices would be different in different states but one state may be more expensive on average than another

TODO - Analysis into the market trends (e.g. 2008 financial crash) and how the model may account for that in future (market trend analysis)

TODO - More important to show in depth analysis i.e. data prep and eda

TODO - Think as the investor in this area

### Problem Statement

Real estate represents a significant portion of most people's wealth, and this is especially true for many homeowners in the United States. A number of factors drive the real estate market including government policies, demographics of the potential buyers,affordability, disparity in housing access, location, cash flows and liquidity as well as the current economic climate. The many variables can make the process tedious for the buyers. Naruto consultants hopes to create a predictive time series model that can help to determine the top five zipcodes in which to invest in.

### Objectives
The main objective of the project is to establish the 5 best zip codes that are the best to invest in.

The objectives of this project are as follows:
 - Identify the key metrics that would be used to classify profitability.
 - Identify the criteria that would classify a house as a "high-end" house or "affordable" house.

The model will be considered a success if it achieves a RMSE of less than 5% and the highest ROI.

### Potential Challenges
- Potentially large amounts of missing data with no conclusive reason, forcing assumptions to be made.
- Some trends may be hard to explain due to differences in locale and perceptions of some matters


## PRELIMINARIES

### Loading of Necessary Modules/Packages

All necessary python packages are loaded at once to make the work easy to view and cleaner as well. 

In [1]:
# Importing necessary libraries
# Basics
import pandas as pd
import numpy as np
import itertools
from io import StringIO

# Visualization
import matplotlib.pyplot as plt
import plotly.express as px
import seaborn as sns
import matplotlib.patches as mpatches
from matplotlib.pylab import rcParams
import time

# Modeling
import statsmodels.api as sm
from statsmodels.graphics.tsaplots import plot_acf, plot_pacf
from statsmodels.tsa.statespace.sarimax import SARIMAX
from sklearn.metrics import mean_squared_error
from statsmodels.tsa.seasonal import seasonal_decompose
from statsmodels.tsa.arima.model import ARIMA
from statsmodels.tsa.stattools import acf, pacf, adfuller
from sklearn.linear_model import LassoLarsCV

# Warnings
import warnings
from statsmodels.tools.sm_exceptions import ConvergenceWarning
warnings.simplefilter('ignore', ConvergenceWarning)
warnings.filterwarnings('ignore')

# Custom Options
pd.set_option('display.max_rows', None)
pd.set_option('display.max_columns',100)


### Definition of Custom Classes & Functions

In line with the concepts of OOP, custom classes were created to speed up the opening and manipulation of data in this project. They are as follows: 

## 2. DATA UNDERSTANDING

### a) Loading & Viewing Data

We start by loading the data as a dataframe.

In [4]:
df = pd.read_csv('./data/zillow_data.csv')

The `df.head()` function is used to get a rough look at a few of the records in the dataframe to understand the data better.

In [5]:
df.head(10)

RegionID  RegionName           City State              Metro  \
0     84654       60657        Chicago    IL            Chicago   
1     90668       75070       McKinney    TX  Dallas-Fort Worth   
2     91982       77494           Katy    TX            Houston   
3     84616       60614        Chicago    IL            Chicago   
4     93144       79936        El Paso    TX            El Paso   
5     91733       77084        Houston    TX            Houston   
6     61807       10467       New York    NY           New York   
7     84640       60640        Chicago    IL            Chicago   
8     91940       77449           Katy    TX            Houston   
9     97564       94109  San Francisco    CA      San Francisco   

      CountyName  SizeRank   1996-04   1996-05   1996-06   1996-07   1996-08  \
0           Cook         1  334200.0  335400.0  336500.0  337600.0  338500.0   
1         Collin         2  235700.0  236900.0  236700.0  235400.0  233300.0   
2         Harris         3  210400.0  212200.0  212200.0  210700.0  208300.0   
3           Cook         4  498100.0  500900.0  503100.0  504600.0  505500.0   
4        El Paso         5   77300.0   77300.0   77300.0   77300.0   77400.0   
5         Harris         6   95000.0   95200.0   95400.0   95700.0   95900.0   
6          Bronx         7  152900.0  152700.0  152600.0  152400.0  152300.0   
7           Cook         8  216500.0  216700.0  216900.0  217000.0  217100.0   
8         Harris         9   95400.0   95600.0   95800.0   96100.0   96400.0   
9  San Francisco        10  766000.0  771100.0  776500.0  781900.0  787300.0   

    1996-09   1996-10   1996-11   1996-12   1997-01   1997-02   1997-03  \
0  339500.0  340400.0  341300.0  342600.0  344400.0  345700.0  346700.0   
1  230600.0  227300.0  223400.0  219600.0  215800.0  211100.0  205700.0   
2  205500.0  202500.0  199800.0  198300.0  197300.0  195400.0  193000.0   
3  505700.0  505300.0  504200.0  503600.0  503400.0  502200.0  500000.0   
4   77500.0   77600.0   77700.0   77700.0   77800.0   77900.0   77900.0   
5   96100.0   96200.0   96100.0   96000.0   95800.0   95500.0   95300.0   
6  152000.0  151800.0  151600.0  151600.0  151700.0  151800.0  151800.0   
7  217200.0  217500.0  217900.0  218600.0  219700.0  220900.0  221800.0   
8   96700.0   96800.0   96800.0   96700.0   96600.0   96400.0   96200.0   
9  793000.0  799100.0  805800.0  814400.0  824300.0  833800.0  842900.0   

    1997-04   1997-05   1997-06   1997-07   1997-08   1997-09   1997-10  \
0  347800.0  349000.0  350400.0  352000.0  353900.0  356200.0  358800.0   
1  200900.0  196800.0  193600.0  191400.0  190400.0  190800.0  192700.0   
2  191800.0  191800.0  193000.0  195200.0  198400.0  202800.0  208000.0   
3  497900.0  496300.0  495200.0  494700.0  494900.0  496200.0  498600.0   
4   77800.0   77800.0   77800.0   77800.0   77800.0   77900.0   78100.0   
5   95100.0   95100.0   95200.0   95400.0   95600.0   95800.0   96000.0   
6  151900.0  152000.0  152200.0  152400.0  152500.0  152600.0  152700.0   
7  223000.0  224200.0  225600.0  227100.0  228800.0  230600.0  232700.0   
8   96100.0   96200.0   96300.0   96600.0   97000.0   97500.0   98000.0   
9  852900.0  863500.0  874800.0  886500.0  898200.0  910200.0  922800.0   

    1997-11   1997-12   1998-01   1998-02    1998-03    1998-04    1998-05  \
0  361800.0  365700.0  370200.0  374700.0   378900.0   383500.0   388300.0   
1  196000.0  201300.0  207400.0  212200.0   214600.0   215100.0   213400.0   
2  213800.0  220700.0  227500.0  231800.0   233400.0   233900.0   233500.0   
3  502000.0  507600.0  514900.0  522200.0   529500.0   537900.0   546900.0   
4   78200.0   78400.0   78600.0   78800.0    79000.0    79100.0    79200.0   
5   96200.0   96400.0   96500.0   96600.0    96600.0    96700.0    96900.0   
6  152900.0  153200.0  153800.0  154300.0   154700.0   155200.0   155700.0   
7  235000.0  237900.0  241300.0  244600.0   247900.0   251300.0   254800.0   
8   98400.0   98800.0   99200.

The `df.info()` and `df.dtypes` functions are both called to give a rough understanding of the dataframe, and the types of data held in each column. Normally, `df.info()` would be sufficient, but due to the sheer number of columns in the dataframe, that information isn't displayed. Thus, `df.dtypes` is called. 

In [6]:
print(df.info())
print()
print(df.dtypes)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 14723 entries, 0 to 14722
Columns: 272 entries, RegionID to 2018-04
dtypes: float64(219), int64(49), object(4)
memory usage: 30.6+ MB
None

RegionID        int64
RegionName      int64
City           object
State          object
Metro          object
CountyName     object
SizeRank        int64
1996-04       float64
1996-05       float64
1996-06       float64
1996-07       float64
1996-08       float64
1996-09       float64
1996-10       float64
1996-11       float64
1996-12       float64
1997-01       float64
1997-02       float64
1997-03       float64
1997-04       float64
1997-05       float64
1997-06       float64
1997-07       float64
1997-08       float64
1997-09       float64
1997-10       float64
1997-11       float64
1997-12       float64
1998-01       float64
1998-02       float64
1998-03       float64
1998-04       float64
1998-05       float64
1998-06       float64
1998-07       float64
1998-08       float64
1998-09       floa

As expected, all the columns representing time-series data will be storing either an integer or a float data type. It is also important to note that the `RegionID` & `RegionName` are stored as integers - to ensure they are unique - and the remaining 5 columns all contain strings.

The shape of the dataframe was observed as follows: 

In [7]:
df.shape

(14723, 272)

The dataframe has 14723 records and 272 column. However, majority of those columns represent the time series values. The first seven columns - `RegionID`, `RegionName`, `City`, `State`, `Metro`, `CountyName` and `SizeRank` - are the columns that give us more information about the dataset. 

Each of these columns is intended to hold a certain type of data as follows: 
- *RegionID*: The unique ID of the region in question.
- *RegionName*: The name of the region in question.
- *City*: The name of the city within a given region.
- *State*: The state in which the RegionID is found.
- *Metro*: The metropolitan name within which the RegionID is found.
- *CountyName*: The name of the county within a given region.
- *SizeRank*: The region's area ranking vis-a-vis other regions, organised in descending order.

Going forward, these columns shall be referred to as the `columns of interest`, while the remaining columns shall be referred to as the `time_series_cols`.

In [8]:
# ['RegionID', 'RegionName', 'City', 'State', 'Metro', 'CountyName','SizeRank']

cols_of_interest = df.columns[:7]
cols_of_interest

Index(['RegionID', 'RegionName', 'City', 'State', 'Metro', 'CountyName',
       'SizeRank'],
      dtype='object')

In [9]:
time_series_cols = df.columns[-265:]
time_series_cols

Index(['1996-04', '1996-05', '1996-06', '1996-07', '1996-08', '1996-09',
       '1996-10', '1996-11', '1996-12', '1997-01',
       ...
       '2017-07', '2017-08', '2017-09', '2017-10', '2017-11', '2017-12',
       '2018-01', '2018-02', '2018-03', '2018-04'],
      dtype='object', length=265)

### b) Data Cleaning

#### **i. Handling of Null Values**

The number of null values in the dataframe will now be established. This is necessary to ensure an appropriate strategy is used when deciding how to handle null values

In [10]:
df.isnull().sum()

RegionID         0
RegionName       0
City             0
State            0
Metro         1043
CountyName       0
SizeRank         0
1996-04       1039
1996-05       1039
1996-06       1039
1996-07       1039
1996-08       1039
1996-09       1039
1996-10       1039
1996-11       1039
1996-12       1039
1997-01       1039
1997-02       1039
1997-03       1039
1997-04       1039
1997-05       1039
1997-06       1039
1997-07       1038
1997-08       1038
1997-09       1038
1997-10       1038
1997-11       1038
1997-12       1038
1998-01       1036
1998-02       1036
1998-03       1036
1998-04       1036
1998-05       1036
1998-06       1036
1998-07       1036
1998-08       1036
1998-09       1036
1998-10       1036
1998-11       1036
1998-12       1036
1999-01       1036
1999-02       1036
1999-03       1036
1999-04       1036
1999-05       1036
1999-06       1036
1999-07       1036
1999-08       1036
1999-09       1036
1999-10       1036
1999-11       1036
1999-12       1036
2000-01     

Aside from the columns that contain time-series values, only the metro column has null values. This can be seen as inconsequential as there is enough data from the other columns to overlook this. 

Taking a closer look at the missing time-series values, there seems to be a steady trend. The values do not appear to be random, as they steadily decrease from April 1996 to July 2014, from where they remain 0 throughout till April 2018.

The gradual change indicates that the presence of null values in these columns is anything but random. As such, when dealing with the null values in the time-series value columns, **all null values will be left as is**. No replacement or removal of records will be done. This is done as it is assumped that not all the houses were built at the same time, thus it is expected that there will be null values for some houses and not others. Furthermore, some of these null values are attributed to the differential times that the houses were put on the market.

When dealing with the null values in the metro column, all null values will be left as the data missing does not impact the objectives of the project. 



#### **ii. Duplicate Values Check**

A check for duplicate values is done. This check specifically applies to the non time-series columns, as it possible to have duplicates in the value of a property, and such duplicates do not serve to understand the dataset better.

In [11]:
duplicates = df.duplicated(subset=cols_of_interest,keep=False)

df[duplicates]

Empty DataFrame
Columns: [RegionID, RegionName, City, State, Metro, CountyName, SizeRank, 1996-04, 1996-05, 1996-06, 1996-07, 1996-08, 1996-09, 1996-10, 1996-11, 1996-12, 1997-01, 1997-02, 1997-03, 1997-04, 1997-05, 1997-06, 1997-07, 1997-08, 1997-09, 1997-10, 1997-11, 1997-12, 1998-01, 1998-02, 1998-03, 1998-04, 1998-05, 1998-06, 1998-07, 1998-08, 1998-09, 1998-10, 1998-11, 1998-12, 1999-01, 1999-02, 1999-03, 1999-04, 1999-05, 1999-06, 1999-07, 1999-08, 1999-09, 1999-10, 1999-11, 1999-12, 2000-01, 2000-02, 2000-03, 2000-04, 2000-05, 2000-06, 2000-07, 2000-08, 2000-09, 2000-10, 2000-11, 2000-12, 2001-01, 2001-02, 2001-03, 2001-04, 2001-05, 2001-06, 2001-07, 2001-08, 2001-09, 2001-10, 2001-11, 2001-12, 2002-01, 2002-02, 2002-03, 2002-04, 2002-05, 2002-06, 2002-07, 2002-08, 2002-09, 2002-10, 2002-11, 2002-12, 2003-01, 2003-02, 2003-03, 2003-04, 2003-05, 2003-06, 2003-07, 2003-08, 2003-09, 2003-10, 2003-11, 2003-12, ...]
Index: []

[0 rows x 272 columns]

No duplicates are observed in the dataset.

#### **iii. Consistent Data Types**

Having run `df.dtypes` earlier, it was noted that some of the price data in the time-series columns was saved as integers while others were saved as floats. All the data in the time-series columns will now be converted to floats since by doing so, we are sure we are not losing/ rounding off any of the price values. 

In [12]:
# Ensuring all time series columns are converted to float64 i.e. the last 265 columns

for col in time_series_cols:
  if df[col].dtype is not 'float64':
    df[col] = df[col].astype('float64')


df.dtypes

RegionID        int64
RegionName      int64
City           object
State          object
Metro          object
CountyName     object
SizeRank        int64
1996-04       float64
1996-05       float64
1996-06       float64
1996-07       float64
1996-08       float64
1996-09       float64
1996-10       float64
1996-11       float64
1996-12       float64
1997-01       float64
1997-02       float64
1997-03       float64
1997-04       float64
1997-05       float64
1997-06       float64
1997-07       float64
1997-08       float64
1997-09       float64
1997-10       float64
1997-11       float64
1997-12       float64
1998-01       float64
1998-02       float64
1998-03       float64
1998-04       float64
1998-05       float64
1998-06       float64
1998-07       float64
1998-08       float64
1998-09       float64
1998-10       float64
1998-11       float64
1998-12       float64
1999-01       float64
1999-02       float64
1999-03       float64
1999-04       float64
1999-05       float64
1999-06   

### c) Data Handling

#### **i. Aggregation of Data**

As all values in the time-series columns are deemed important, aggregating the data to a yearly basis, for example, at such an early stage was deemed unnecessary. This is because a lot of vital information would be lost in the process that would be necessary later on. Thus, aggregation would be done on an as-needed basis.

#### **ii. Grouping of Data**

Data will now be grouped by state, to give a better idea as to which states have more property listings than others:

In [13]:
df[['State', 'RegionID','RegionName','City','Metro','CountyName','SizeRank']].groupby("State").count()

RegionID  RegionName  City  Metro  CountyName  SizeRank
State                                                         
AK           28          28    28     22          28        28
AL          183         183   183    175         183       183
AR          105         105   105    100         105       105
AZ          230         230   230    228         230       230
CA         1224        1224  1224   1182        1224      1224
CO          249         249   249    229         249       249
CT          124         124   124    124         124       124
DC           18          18    18     18          18        18
DE           41          41    41     41          41        41
FL          785         785   785    758         785       785
GA          345         345   345    317         345       345
HI           62          62    62     62          62        62
IA          158         158   158    156         158       158
ID          110         110   110     87         110       110
IL          547         547   547    537         547       547
IN          428         428   428    363         428       428
KS          241         241   241    184         241       241
KY          139         139   139    132         139       139
LA          193         193   193    173         193       193
MA          417         417   417    416         417       417
MD          317         317   317    300         317       317
ME           86          86    86     85          86        86
MI          499         499   499    457         499       499
MN          375         375   375    339         375       375
MO          319         319   319    269         319       319
MS          153         153   153     98         153       153
MT           71          71    71     43          71        71
NC          428         428   428    391         428       428
ND           31          31    31     21          31        31
NE           83          83    83     78          83        83
NH          199         199   199    178         199       199
NJ          502         502   502    502         502       502
NM           60          60    60     55          60        60
NV          103         103   103    103         103       103
NY         1015        1015  1015    963        1015      1015
OH          588         588   588    570         588       588
OK          221         221   221    200         221       221
OR          224         224   224    216         224       224
PA          831         831   831    816         831       831
RI           59          59    59     59          59        59
SC          206         206   206    199         206       206
SD           19          19    19     19          19        19
TN          404         404   404    336         404       404
TX          989         989   989    873         989       989
UT          121         121   121    109         121       121
VA          401         401   401    365         401       401
VT           16          16    16     16          16        16
WA          341         341   341    323         341       341
WI          332         332   332    302         332       332
WV           72          72    72     71          72        72
WY           31          31    31     20          31        31

Particular focus will be placed on the `RegionID` as that column does not have any null values, thus an accurate count of properties can be done.

In [14]:
df_state_grouping = df[['State', 'RegionID']].groupby(
    "State").count().sort_values(by='RegionID', ascending=False)
df_state_grouping

RegionID
State          
CA         1224
NY         1015
TX          989
PA          831
FL          785
OH          588
IL          547
NJ          502
MI          499
IN          428
NC          428
MA          417
TN          404
VA          401
MN          375
GA          345
WA          341
WI          332
MO          319
MD          317
CO          249
KS          241
AZ          230
OR          224
OK          221
SC          206
NH          199
LA          193
AL          183
IA          158
MS          153
KY          139
CT          124
UT          121
ID          110
AR          105
NV          103
ME           86
NE           83
WV           72
MT           71
HI           62
NM           60
RI           59
DE           41
WY           31
ND           31
AK           28
SD           19
DC           18
VT           16

California (CA) is seen to have the most listings with 1224, followed closely by New York, NY (1015) and Texas, TX (989). Vermont (VT), Washington DC (DC) and San Diego (SD) have the least listings at 16, 18 and 19 respectively. 

#### **iii. Presence of Outliers** 

There will be outliers present in the data, that will be more apparent in the EDA section. These outliers may come around as a result of over-inflated house prices, or simply having houses marked below their market value by owners who are in a rush to sell. Whatever the case may be, the outliers in the dataset **will be kept throughout the analysis**. 

These outliers play an important role in developing a more realistic model, as the world of real-estate will always have overpriced or underpriced properties that do not conform to the norms. Keeping these outliers in will serve to create a more realistic model.

### d) Feature Engineering

#### **i. Addition of New Features** 

Relevant features that could affect house prices were noted and their creation was deemed necessary. These features are: 

- *Return On Investment(ROI)(%)*: Calculated as 
$
\left( \frac{{\text{{Last Price of Property}}}}{{\text{{Initial Price of Property}}}} - 1 \right) \times 100\% 
$

As we know that the latter, and in particular the last time-series column, has no null values, this column will automatically be our numerator (i.e. `df.columns[271]`). As for the denominator, it's value will be occuppied by the oldest datetime that is not a null value. Therefore, the calculation of the ROI was done as follows:

In [17]:
for index, row in df.iterrows():
  # Establishing the last price of the property
  last_price = df.columns[271]

  # Establishing the initial price of the property
  initial_price = 0

  # Loop through time_series_cols to find the first non-null value
  for col in time_series_cols:
      initial_price = df[col].values[index]
      if not pd.isnull(initial_price):
        # print(f"Record {index} initial price: {initial_price} @ {col}")
        break
  df.at[index, 'Initial Price'] = initial_price
  
  df['Last Price'] = df[last_price]
  
  df['ROI (%)'] = round(
      (((df['Last Price'] / df['Initial Price']) - 1) * 100), 2)

df.head(10)

RegionID  RegionName           City State              Metro  \
0     84654       60657        Chicago    IL            Chicago   
1     90668       75070       McKinney    TX  Dallas-Fort Worth   
2     91982       77494           Katy    TX            Houston   
3     84616       60614        Chicago    IL            Chicago   
4     93144       79936        El Paso    TX            El Paso   
5     91733       77084        Houston    TX            Houston   
6     61807       10467       New York    NY           New York   
7     84640       60640        Chicago    IL            Chicago   
8     91940       77449           Katy    TX            Houston   
9     97564       94109  San Francisco    CA      San Francisco   

      CountyName  SizeRank   1996-04   1996-05   1996-06   1996-07   1996-08  \
0           Cook         1  334200.0  335400.0  336500.0  337600.0  338500.0   
1         Collin         2  235700.0  236900.0  236700.0  235400.0  233300.0   
2         Harris         3  210400.0  212200.0  212200.0  210700.0  208300.0   
3           Cook         4  498100.0  500900.0  503100.0  504600.0  505500.0   
4        El Paso         5   77300.0   77300.0   77300.0   77300.0   77400.0   
5         Harris         6   95000.0   95200.0   95400.0   95700.0   95900.0   
6          Bronx         7  152900.0  152700.0  152600.0  152400.0  152300.0   
7           Cook         8  216500.0  216700.0  216900.0  217000.0  217100.0   
8         Harris         9   95400.0   95600.0   95800.0   96100.0   96400.0   
9  San Francisco        10  766000.0  771100.0  776500.0  781900.0  787300.0   

    1996-09   1996-10   1996-11   1996-12   1997-01   1997-02   1997-03  \
0  339500.0  340400.0  341300.0  342600.0  344400.0  345700.0  346700.0   
1  230600.0  227300.0  223400.0  219600.0  215800.0  211100.0  205700.0   
2  205500.0  202500.0  199800.0  198300.0  197300.0  195400.0  193000.0   
3  505700.0  505300.0  504200.0  503600.0  503400.0  502200.0  500000.0   
4   77500.0   77600.0   77700.0   77700.0   77800.0   77900.0   77900.0   
5   96100.0   96200.0   96100.0   96000.0   95800.0   95500.0   95300.0   
6  152000.0  151800.0  151600.0  151600.0  151700.0  151800.0  151800.0   
7  217200.0  217500.0  217900.0  218600.0  219700.0  220900.0  221800.0   
8   96700.0   96800.0   96800.0   96700.0   96600.0   96400.0   96200.0   
9  793000.0  799100.0  805800.0  814400.0  824300.0  833800.0  842900.0   

    1997-04   1997-05   1997-06   1997-07   1997-08   1997-09   1997-10  \
0  347800.0  349000.0  350400.0  352000.0  353900.0  356200.0  358800.0   
1  200900.0  196800.0  193600.0  191400.0  190400.0  190800.0  192700.0   
2  191800.0  191800.0  193000.0  195200.0  198400.0  202800.0  208000.0   
3  497900.0  496300.0  495200.0  494700.0  494900.0  496200.0  498600.0   
4   77800.0   77800.0   77800.0   77800.0   77800.0   77900.0   78100.0   
5   95100.0   95100.0   95200.0   95400.0   95600.0   95800.0   96000.0   
6  151900.0  152000.0  152200.0  152400.0  152500.0  152600.0  152700.0   
7  223000.0  224200.0  225600.0  227100.0  228800.0  230600.0  232700.0   
8   96100.0   96200.0   96300.0   96600.0   97000.0   97500.0   98000.0   
9  852900.0  863500.0  874800.0  886500.0  898200.0  910200.0  922800.0   

    1997-11   1997-12   1998-01   1998-02    1998-03    1998-04    1998-05  \
0  361800.0  365700.0  370200.0  374700.0   378900.0   383500.0   388300.0   
1  196000.0  201300.0  207400.0  212200.0   214600.0   215100.0   213400.0   
2  213800.0  220700.0  227500.0  231800.0   233400.0   233900.0   233500.0   
3  502000.0  507600.0  514900.0  522200.0   529500.0   537900.0   546900.0   
4   78200.0   78400.0   78600.0   78800.0    79000.0    79100.0    79200.0   
5   96200.0   96400.0   96500.0   96600.0    96600.0    96700.0    96900.0   
6  152900.0  153200.0  153800.0  154300.0   154700.0   155200.0   155700.0   
7  235000.0  237900.0  241300.0  244600.0   247900.0   251300.0   254800.0   
8   98400.0   98800.0   99200.

#### **ii. Normalization/ Scaling**

Having gone through the data, no different scales are noticed. However, there may be a need for normalization when it comes to comparing regions with vastly different quantities of properties listed that have a variety of prices. This is to ensure that the comparison is as accurate as possible and without an inherent bias or skew. This shall be done later on in the EDA stages on an as-needed basis

#### **iii. Handling ** 

A new dataframe is now created which contains the melted data. This will convert the original dataframe `df` from wide to long format. This will be done by applying the custom function `melt_data`

In [ ]:
def melt_data(df):
    """
    Takes the zillow_data dataset in wide form or a subset of the zillow_dataset.  
    Returns a long-form datetime dataframe 
    with the datetime column names as the index and the values as the 'values' column.
    
    If more than one row is passes in the wide-form dataset, the values column
    will be the mean of the values from the datetime columns in all of the rows.
    """

    melted = pd.melt(df, id_vars=['RegionName', 'RegionID', 'SizeRank',
                     'City', 'State', 'Metro', 'CountyName'], var_name='time')
    melted['time'] = pd.to_datetime(melted['time'], infer_datetime_format=True)
    melted = melted.dropna(subset=['value'])
    return melted.groupby('time').aggregate({'value': 'mean'})

Converting the dataframe to long format:

In [ ]:
df_melted = melt_data(df)

## 3. EDA & VISUALISATION

### Descriptive Statistics

The mean, median & standard deviation (as well as other statistics) are seen here: 

In [18]:
df.describe()

RegionID    RegionName      SizeRank       1996-04       1996-05  \
count   14723.000000  14723.000000  14723.000000  1.368400e+04  1.368400e+04   
mean    81075.010052  48222.348706   7362.000000  1.182991e+05  1.184190e+05   
std     31934.118525  29359.325439   4250.308342  8.600251e+04  8.615567e+04   
min     58196.000000   1001.000000      1.000000  1.130000e+04  1.150000e+04   
25%     67174.500000  22101.500000   3681.500000  6.880000e+04  6.890000e+04   
50%     78007.000000  46106.000000   7362.000000  9.950000e+04  9.950000e+04   
75%     90920.500000  75205.500000  11042.500000  1.432000e+05  1.433000e+05   
max    753844.000000  99901.000000  14723.000000  3.676700e+06  3.704200e+06   

            1996-06       1996-07       1996-08       1996-09       1996-10  \
count  1.368400e+04  1.368400e+04  1.368400e+04  1.368400e+04  1.368400e+04   
mean   1.185374e+05  1.186531e+05  1.187803e+05  1.189275e+05  1.191205e+05   
std    8.630923e+04  8.646795e+04  8.665094e+04  8.687208e+04  8.715185e+04   
min    1.160000e+04  1.180000e+04  1.180000e+04  1.200000e+04  1.210000e+04   
25%    6.910000e+04  6.920000e+04  6.937500e+04  6.950000e+04  6.960000e+04   
50%    9.970000e+04  9.970000e+04  9.980000e+04  9.990000e+04  9.995000e+04   
75%    1.432250e+05  1.432250e+05  1.435000e+05  1.437000e+05  1.439000e+05   
max    3.729600e+06  3.754600e+06  3.781800e+06  3.813500e+06  3.849600e+06   

            1996-11       1996-12       1997-01       1997-02       1997-03  \
count  1.368400e+04  1.368400e+04  1.368400e+04  1.368400e+04  1.368400e+04   
mean   1.193454e+05  1.196851e+05  1.201213e+05  1.205442e+05  1.208966e+05   
std    8.747981e+04  8.791269e+04  8.840935e+04  8.887373e+04  8.927704e+04   
min    1.220000e+04  1.230000e+04  1.250000e+04  1.260000e+04  1.270000e+04   
25%    6.980000e+04  7.000000e+04  7.030000e+04  7.060000e+04  7.090000e+04   
50%    1.001000e+05  1.003000e+05  1.006000e+05  1.010000e+05  1.013000e+05   
75%    1.441250e+05  1.443000e+05  1.446000e+05  1.451000e+05  1.455000e+05   
max    3.888900e+06  3.928800e+06  3.964600e+06  3.995600e+06  4.023800e+06   

            1997-04       1997-05       1997-06       1997-07       1997-08  \
count  1.368400e+04  1.368400e+04  1.368400e+04  1.368500e+04  1.368500e+04   
mean   1.212682e+05  1.216559e+05  1.220531e+05  1.224563e+05  1.228799e+05   
std    8.969700e+04  9.012583e+04  9.056147e+04  9.100389e+04  9.146850e+04   
min    1.290000e+04  1.300000e+04  1.300000e+04  1.320000e+04  1.330000e+04   
25%    7.110000e+04  7.130000e+04  7.170000e+04  7.200000e+04  7.230000e+04   
50%    1.015000e+05  1.017000e+05  1.019000e+05  1.021000e+05  1.024000e+05   
75%    1.457000e+05  1.459250e+05  1.462000e+05  1.466000e+05  1.469000e+05   
max    4.048300e+06  4.066900e+06  4.080400e+06  4.090800e+06  4.101500e+06   

            1997-09       1997-10       1997-11       1997-12       1998-01  \
count  1.368500e+04  1.368500e+04  1.368500e+04  1.368500e+04  1.368700e+04   
mean   1.233274e+05  1.238231e+05  1.243537e+05  1.250026e+05  1.257367e+05   
std    9.196320e+04  9.250608e+04  9.308367e+04  9.374830e+04  9.445247e+04   
min    1.350000e+04  1.350000e+04  1.360000e+04  1.380000e+04  1.400000e+04   
25%    7.240000e+04  7.280000e+04  7.310000e+04  7.350000e+04  7.400000e+04   
50%    1.026000e+05  1.030000e+05  1.034000e+05  1.039000e+05  1.044000e+05   
75%    1.472000e+05  1.477000e+05  1.482000e+05  1.489000e+05  1.494500e+05   
max    4.115200e+06  4.131900e+06  4.150200e+06  4.165900e+06  4.175400e+06   

            1998-02       1998-03       1998-04       1998-05       1998-06  \
count  1.368700e+04  1.368700e+04  1.368700e+04  1.368700e+04  1.368700e+04   
mean   1.264508e+05  1.270863e+05  1.277364e+05  1.283850e+05  1.290327e+05   
std    9.510499e+04  9.567485e+04  9.624843e+04  9.681400e+04  9.737949e+04   
min    1.410000e+04  1.420000e+04  1.440000e+04  1.450000e+04  1.460000e+04   
25%    7.440000e+04  7.470000e+04  7.520000e+04 

## 4. ARIMA MODELLING

## 5. FINE TUNING WITH PROPHET

## 6. INTERPRETING RESULTS